### PR TITLE
sys/auto_init: custom auto-initialization sequence (simple)

### PR DESCRIPTION
**Pull request by fabian18**
### Contribution description

This is another proposal of how to utilize `XFA`s to build a custom auto-init sequence of modules.
Priorities are statically assigned numbers for now.
You can chose whether to use the current sequence or switch on `CONFIG_AUTO_INIT_PRIO_EXPERIMENTAL` to enable a more group oriented sequence.  If we want to include this option, we can still discuss. I think it does not harm and it moves us a little forward on #16529.

### Testing procedure

`CFLAGS+="-DCONFIG_AUTO_INIT_ENABLE_DEBUG=1 -DCONFIG_AUTO_INIT_PRIO_EXPERIMENTAL=1" BOARD=nucleo-f767zi make flash term`

```
2022-03-11 10:52:31,975 # ��auto_init ztimer (3010): 
2022-03-11 10:52:31,976 # auto_init random (3040): 
2022-03-11 10:52:31,979 # auto_init gnrc_netif (4050): 
2022-03-11 10:52:31,983 # NETOPT_RX_END_IRQ not implemented by driver
2022-03-11 10:52:31,987 # NETOPT_TX_END_IRQ not implemented by driver
2022-03-11 10:52:31,990 # auto_init gnrc_pktbuf (6010): 
2022-03-11 10:52:31,993 # auto_init gnrc_ipv6_nib (6080): 
2022-03-11 10:52:31,996 # auto_init gnrc_ipv6 (6090): 
2022-03-11 10:52:31,998 # auto_init gnrc_rpl (6100): 
2022-03-11 10:52:32,001 # auto_init gnrc_udp (6110): 
2022-03-11 10:52:32,004 # auto_init gnrc_pktdump (7010): 
2022-03-11 10:52:32,011 # main(): This is RIOT! (Version: 2022.04-devel-660-g445e89-auto_init_using_xfa_simple)
2022-03-11 10:52:32,015 # RIOT network stack example application
2022-03-11 10:52:32,017 # All up, running the shell now
```

### Issues/PRs references

Substitutes #17109

### CODE REVIEW

**Comment by kaspar030** [comment]
> You can chose whether to use the current sequence or switch on `CONFIG_AUTO_INIT_PRIO_EXPERIMENTAL` to enable a more group oriented sequence. If we want to include this option, we can still discuss. I think it does not harm and it moves us a little forward on #16529.

May I suggest to start with "legacy", once in, just change to the new? There's no need to have both in-tree.
A follow-up re-grouping the sequences could then be tested and reviewed independently.

File: sys/include/auto_init_utils.h [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+/**
+ * @brief   Add a module to the auto-initialization array
+ *
+ * @param   module      Module name
+ * @param   prio        Priority level
+ * @param   function    Function to be called on initialization
+ * @param   ...         String format and arguments to be printed with @ref DEBUG()
+ */
+#define AUTO_INIT_MODULE(module, prio, function, ...)                                   \
+    static inline void CONCAT(CONCAT(_debug_, prio), function)(void) {                  \
```
Comment (kaspar030) [inline sys/include/auto_init_utils.h:107]: To my taste, there's too much macro magic. A macro-generated function calling out to printf? Not a fan.

How about we add fields to auto_init_module_t, like, module_name, prio, extra. "extra" could be a user supplied `const char *` (string literal), so multiple instances of the same module can be differentiated.
The printing can then be done in plain C in `auto_init()`.

That way we could drop *all* macros from this file but `AUTO_INIT_MODULE` (assuming XTSTR is re-used from macros/xtstr.h, for setting the module name field from "module" arg).

**Review by kaspar030 — CHANGES_REQUESTED** [review]
I like this much better than the previous version!

File: sys/include/auto_init_utils.h (hunk 1 above)
Comment (fabian18) [reply sys/include/auto_init_utils.h:107]: Ok, but do we always want to pay for these struct members. I guess only if debug is enabled. Or should each member get its own Kconfig flag? Like `CONFIG_AUTO_INIT_MODULE_HAS_NAME`, `CONFIG_AUTO_INIT_MODULE_HAS_PRIO`, `CONFIG_AUTO_INIT_MODULE_HAS_EXTRA`.

Comment (kaspar030) [reply sys/include/auto_init_utils.h:107]: yeah, `#ifdef AUTOINIT_DEBUG` (or whatever) only

File: sys/include/auto_init_utils.h [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+ * @brief   Add a module to the auto-initialization array
+ *
+ * @param   module      Module name
+ * @param   priority    Priority level
+ * @param   function    Function to be called on initialization
+ * @param   ...         Optional debug string
+ */
+#define AUTO_INIT_MODULE(module, priority, function, ...)                               \
+    XFA_CONST(auto_init_xfa, priority)                                                  \
+    auto_init_module_t auto_init_xfa_ ## module                                         \
```
Comment (kaspar030) [inline sys/include/auto_init_utils.h:79]: Why not fold thise two if cases? easier to spot the differences.

I mean this:
```
auto_init_module_t auto_init_xfa_ ## module                                         \
= (auto_init_module_t) { (auto_init_fn_t)function,                              \
#if CONFIG_AUTO_INIT_ENABLE_DEBUG \
 priority, XTSTR(module), ""__VA_ARGS__ \
#endif
}
```

Comment (fabian18) [reply sys/include/auto_init_utils.h:79]: If you mean this:
```C
#define AUTO_INIT_MODULE(module, priority, function, ...)                               \
    XFA_CONST(auto_init_xfa, priority)                                                  \
    auto_init_module_t auto_init_xfa_ ## module                                         \
        = (auto_init_module_t) { (auto_init_fn_t)function,                              \
#if IS_ACTIVE(CONFIG_AUTO_INIT_ENABLE_DEBUG)                                            \
                                 priority, XTSTR(module), ""__VA_ARGS__                 \
#endif                                                                                  \
        }
```
it is not allowed to have an `#if` within a `#define`, unfortunately.

```
RIOT/sys/include/auto_init_utils.h:91:57: error: '#' is not followed by a macro parameter
   91 | #define AUTO_INIT_MODULE(module, priority, function, ...)
```

**Comment by fabian18** [comment]
Can I maybe squash and rebase? There is a new `MODULE_AUTO_INIT_SOCK_DNS`.

**Comment by kaspar030** [comment]
yes, please squash!

File: sys/auto_init/Kconfig [hunk 3]
```diff
@@ -16,6 +16,10 @@ menuconfig MODULE_AUTO_INIT
 
 if MODULE_AUTO_INIT
 
+config AUTO_INIT_ENABLE_DEBUG
+    bool "Print a debug message befor a module is initialized"
```
Comment (kaspar030) [inline sys/auto_init/Kconfig:20]:
```suggestion
    bool "Print a debug message before a module is initialized"
```

Comment (fabian18) [reply sys/auto_init/Kconfig]: Thanks. I squashed this right in.

File: sys/include/auto_init_utils.h [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        = (auto_init_module_t) { (auto_init_fn_t)function,                              \
+                                 priority, XTSTR(module), ""__VA_ARGS__ }
+#else
+#define AUTO_INIT_MODULE(module, priority, function, ...)                               \
+    XFA_CONST(auto_init_xfa, priority)                                                  \
+    auto_init_module_t auto_init_xfa_ ## module                                         \
+        = (auto_init_module_t) { (auto_init_fn_t)function }
```
Comment (fjmolinas) [inline sys/include/auto_init_utils.h:86]:
```suggestion
       = { .init = (auto_init_fn_t)function }
```

File: sys/include/auto_init_utils.h [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+ * @param   module      Module name
+ * @param   priority    Priority level
+ * @param   function    Function to be called on initialization
+ * @param   ...         Optional debug string
+ */
+#define AUTO_INIT_MODULE(module, priority, function, ...)                               \
+    XFA_CONST(auto_init_xfa, priority)                                                  \
+    auto_init_module_t auto_init_xfa_ ## module                                         \
+        = (auto_init_module_t) { (auto_init_fn_t)function,                              \
+                                 priority, XTSTR(module), ""__VA_ARGS__ }
```
Comment (fjmolinas) [inline sys/include/auto_init_utils.h:81]:
```suggestion
        = { .init = (auto_init_fn_t)function,                                           \
            .prio = priority,                                                           \
            .name = XTSTR(module),                                                      \
            .extra = ""__VA_ARGS__ }
```

I don't see why the case is needed, and it actually makes the compilation fail on native.

**Review by fjmolinas — COMMENTED** [review]
Not sure why the cast is needed, but it makes the compilation fail on native, if you think my suggested patch is OK, please squash right away!

**Comment by fjmolinas** [comment]
@kaspar030 have all your change requests been addressed?

**Comment by fabian18** [comment]
> but it makes the compilation fail on native
```
/data/riotbuild/riotbase/sys/auto_init/auto_init.c:62:34: error: ISO C99 requires at least one argument for the "..." in a variadic macro [-Werror]
                  auto_init_random);
```

Oh, `native` also does not like empty `...`.
Beyond suppression, I believe there is no nice workaround for this. And the strategy to handle macros with different numbers of arguments is too much headache for a user supplied debug message.
So are we going to drop the `extra`?

**Comment by jue89** [comment]
> So are we going to drop the extra?

I'd suggest: yes. Keep it stupid simple ;-)

**Comment by kaspar030** [comment]
does it make sense to support an argument (void*)? I could see, e.g., a network driver use that to pass parameters.
*maybe* not worth the extra 4 bytes on *every* auto init function.

**Comment by kaspar030** [comment]
I think we're converging to just `AUTO_INIT(function, priority)`.

**Comment by jue89** [comment]
> does it make sense to support an argument (void*)? I could see, e.g., a network driver use that to pass parameters. _maybe_ not worth the extra 4 bytes on _every_ auto init function.

Nothing should stop us from introducing `AUTO_INIT_WITH_ARG(function, arg, priority)` later, if we encounter cases that require args.

**Comment by fjmolinas** [comment]
The failing test is unrelated to this PR as its already failing in master, bisecting leads to:

```
47dd3b1889c813550c1029d03c06c404a29d42cb is the first bad commit
commit 47dd3b1889c813550c1029d03c06c404a29d42cb
Author: Kaspar Schleiser <kaspar@schleiser.de>
Date:   Thu Aug 6 13:27:13 2020 +0200

    tests: make test_utils_print_stack_usage a default module

 tests/Makefile.tests_common | 2 ++
 1 file changed, 2 insertions(+)
```

**Review by fjmolinas — APPROVED** [review]
ACK, failing test is unrelated to this PR and failing in master. Awesome to see this get in!

**Comment by fjmolinas** [comment]
@kaspar030 if all your change requests have been addressed I woudl trigger the ci again without the tests or on a fast build.

**Review by kaspar030 — APPROVED** [review]

**Comment by fjmolinas** [comment]
ping @OlegHahm

**Comment by OlegHahm** [comment]
Sorry, apparently I have to check my Github notification settings...

I think it would be very nice to have this auto init cleanup in for the release and since I missed the soft feature freeze deadline myself..let's go!

**Comment by fjmolinas** [comment]
Awesome!! Super cool to see this one in! Thanks for sticking through with ith @fabian18!!

**Comment by fabian18** [comment]
Thank you all for reviewing!

**Pull request merged**